### DISCUSSION TITLE
support systemd socket activation

### DISCUSSION THREAD

**Discussion by frankli0324** [outline/outline#3884] [Ideas]
systemd supports opening a placeholder socket and listen for incoming requests. when a request triggers that socket, systemd passes the file descriptor(no matter its a regular tcp port or ipc file socket) to the application as fd 3, and notifies the application with `LISTEN_FDS` and `LISTEN_PID` env variable.

I deploy outline instances on systemd, along with many other services. socket activation helps me keep the service up and running whenever there's a request, without manually starting the service. this also can be used for lazy starting.

**Comment by frankli0324** [comment]
documentation for this can be found at https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html

**Comment by frankli0324** [comment]
https://github.com/nodejs/help/issues/2115

https://github.com/nodejs/node/issues/44035

**Comment by tommoor** [comment]
I don't think this is a common enough usecase to bake into the product, this is literally the first time it's been mentioned in 6 years. I'm still not really sure what the value is adding to be honest, this is a web service – it should always be running on the host machine.

**Reply by frankli0324** [reply]
> it should always be running on the host machine.

I agree with the first part, maybe docker *is* enough for most self-hosting users... but I dont understand what this line is about...

my deployment looks like this:
I deployed outline without docker for a tidy and consistent environment (~~almost~~ all services I use runs as a systemd service). and I don't like outline or any web app taking a tcp port.
I have nginx in charge of terminating tls and proxying the request to apps:
`proxy_pass http://unix:/run/outline/outline.sock`
and systemd in charge of activating outline in case it stops with socket activation.

**Reply by frankli0324** [reply]
in general, socket activation provides service availability.
I'm not sure if outline could also support directly listening on unix sockets for it provides [better performance](https://stackoverflow.com/a/29436429).
